In [513]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

importing the files

In [514]:
pd.set_option("display.max_columns", None) #i wanna see all the columns :D

In [515]:
Products = pd.read_csv('Products.csv')
Customers = pd.read_json('customers.json')
Internet_Sales  = pd.read_excel('InternetSales.xlsx', engine='calamine')
Sales_Territory  = pd.read_parquet('SalesTerritory.parquet')

In [516]:
def bold_green(text):
    return f'\033[1;32m{text}\033[0m'
def bold(text):
    return f'\033[1m{text}\033[0m'

In [517]:
datasets = {
    "Products (CSV)": Products,
    "Customers (JSON)": Customers,
    "InternetSales (xlsx)":Internet_Sales,
    "SalesTerritory":Sales_Territory,
}

In [518]:
for n, d in datasets.items():
    print(f"name: {bold_green(n)}:\n{d.columns}")

name: Products (CSV):
Index(['ProductKey', 'ProductAlternateKey', 'ProductSubcategoryKey',
       'WeightUnitMeasureCode', 'SizeUnitMeasureCode', 'EnglishProductName',
       'SpanishProductName', 'FrenchProductName', 'StandardCost',
       'FinishedGoodsFlag', 'Color', 'SafetyStockLevel', 'ReorderPoint',
       'ListPrice', 'Size', 'SizeRange', 'Weight', 'DaysToManufacture',
       'ProductLine', 'DealerPrice', 'Class', 'Style', 'ModelName',
       'EnglishDescription', 'FrenchDescription', 'ChineseDescription',
       'ArabicDescription', 'HebrewDescription', 'ThaiDescription',
       'GermanDescription', 'JapaneseDescription', 'TurkishDescription',
       'StartDate', 'EndDate', 'Status'],
      dtype='str')
name: Customers (JSON):
Index(['CustomerKey', 'GeographyKey', 'CustomerAlternateKey', 'Title',
       'FirstName', 'MiddleName', 'LastName', 'NameStyle', 'BirthDate',
       'MaritalStatus', 'Suffix', 'Gender', 'EmailAddress', 'YearlyIncome',
       'TotalChildren', 'NumberChild

In [519]:
##datasets infos
for name, data in datasets.items():
    na_ratio = (data.isna().sum() / len(data)) * 100
    dup_ratio = (data.duplicated().sum() / len(data)) * 100
    #print(data.shape)
    print(f'{bold('Dataset')}: {bold_green(name)}:')
    print(f"{bold('Lenght')}: {len(data)}")
    print(f"{bold('Number of columns')}: {len(data.columns)}")
    print(f"{bold('Missing values %')}:\n{(na_ratio[na_ratio > 0])}\n")
    print(f"{bold('Duplicated values %')}:\n{(dup_ratio[dup_ratio > 0])}\n")

Dataset: Products (CSV):
Lenght: 606
Number of columns: 35
Missing values %:
ProductSubcategoryKey    34.488449
WeightUnitMeasureCode    53.465347
SizeUnitMeasureCode      58.250825
SpanishProductName       47.359736
FrenchProductName        47.359736
StandardCost             34.818482
Color                    41.914191
ListPrice                34.818482
Size                     50.660066
SizeRange                50.660066
Weight                   53.465347
ProductLine              37.293729
DealerPrice              34.818482
Class                    45.544554
Style                    50.330033
ModelName                34.488449
EnglishDescription       34.653465
FrenchDescription        34.653465
ChineseDescription       34.653465
ArabicDescription        34.653465
HebrewDescription        34.653465
ThaiDescription          34.653465
GermanDescription        34.653465
JapaneseDescription      34.653465
TurkishDescription       34.653465
EndDate                  66.996700
Status       

In [520]:
for name, data in datasets.items():
    print(f'{bold_green(name)}:')
    print(data.info())

Products (CSV):
<class 'pandas.DataFrame'>
RangeIndex: 606 entries, 0 to 605
Data columns (total 35 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   ProductKey             606 non-null    int64  
 1   ProductAlternateKey    606 non-null    str    
 2   ProductSubcategoryKey  397 non-null    float64
 3   WeightUnitMeasureCode  282 non-null    str    
 4   SizeUnitMeasureCode    253 non-null    str    
 5   EnglishProductName     606 non-null    str    
 6   SpanishProductName     319 non-null    str    
 7   FrenchProductName      319 non-null    str    
 8   StandardCost           395 non-null    float64
 9   FinishedGoodsFlag      606 non-null    int64  
 10  Color                  352 non-null    str    
 11  SafetyStockLevel       606 non-null    int64  
 12  ReorderPoint           606 non-null    int64  
 13  ListPrice              395 non-null    float64
 14  Size                   299 non-null    str    
 15  S

In [521]:
Products.head()

,ProductKey,ProductAlternateKey,ProductSubcategoryKey,WeightUnitMeasureCode,SizeUnitMeasureCode,EnglishProductName,SpanishProductName,FrenchProductName,StandardCost,FinishedGoodsFlag,Color,SafetyStockLevel,ReorderPoint,ListPrice,Size,SizeRange,Weight,DaysToManufacture,ProductLine,DealerPrice,Class,Style,ModelName,EnglishDescription,FrenchDescription,ChineseDescription,ArabicDescription,HebrewDescription,ThaiDescription,GermanDescription,JapaneseDescription,TurkishDescription,StartDate,EndDate,Status
0,1,AR-5381,NaN,NaN,NaN,Adjustable Race,NaN,NaN,NaN,0,NaN,1000,750,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003-07-01,NaN,Current
1,2,BA-8327,NaN,NaN,NaN,Bearing Ball,NaN,NaN,NaN,0,NaN,1000,750,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003-07-01,NaN,Current
2,3,BE-2349,NaN,NaN,NaN,BB Ball Bearing,NaN,NaN,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003-07-01,NaN,Current
3,4,BE-2908,NaN,NaN,NaN,Headset Ball Bearings,NaN,NaN,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003-07-01,NaN,Current
4,5,BL-2036,NaN,NaN,NaN,Blade,NaN,NaN,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,2003-07-01,NaN,Current


In [522]:
Internet_Sales.head()

,ProductKey,OrderDate,ShipDate,CustomerKey,PromotionKey,SalesTerritoryKey,SalesOrderNumber,SalesOrderLineNumber,RevisionNumber,OrderQuantity,UnitPrice,ExtendedAmount,UnitPriceDiscountPct,DiscountAmount,ProductStandardCost,TotalProductCost,29,6,98,19,36,100
0,310,2010-12-29,2011-01-05,21768,1,6,SO43697,1,1,1,3578.2700,3578.2700,0,0,2171.2942,2171.2942,NaN,NaN,NaN,3578.27,NaN,NaN
1,346,2010-12-29,2011-01-05,28389,1,7,SO43698,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,NaN,NaN,NaN,NaN,NaN,NaN
2,346,2010-12-29,2011-01-05,25863,1,1,SO43699,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,NaN,NaN,NaN,NaN,NaN,3399.9900
3,336,2010-12-29,2011-01-05,14501,1,4,SO43700,1,1,1,699.0982,699.0982,0,0,413.1463,413.1463,NaN,NaN,NaN,NaN,NaN,699.0982
4,346,2010-12-29,2011-01-05,11003,1,9,SO43701,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,NaN,3399.99,NaN,NaN,NaN,NaN


In [523]:
Customers.head()

,CustomerKey,GeographyKey,CustomerAlternateKey,Title,FirstName,MiddleName,LastName,NameStyle,BirthDate,MaritalStatus,Suffix,Gender,EmailAddress,YearlyIncome,TotalChildren,NumberChildrenAtHome,EnglishEducation,SpanishEducation,FrenchEducation,EnglishOccupation,SpanishOccupation,FrenchOccupation,HouseOwnerFlag,NumberCarsOwned,AddressLine1,AddressLine2,Phone,DateFirstPurchase,CommuteDistance
0,12725,171,AW00012725,NaN,Gabrielle,NaN,James,0,1916-02-10,M,NaN,F,gabrielle17@adventure-works.com,10000,4,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,0,2,Nonnendamm 6500,NaN,1 (11) 500 555-0125,2013-08-18,1-2 Miles
1,16455,208,AW00016455,NaN,Deb,A,Moreno,0,1917-02-09,M,NaN,F,deb3@adventure-works.com,10000,4,0,High School,Educación secundaria,Bac + 2,Manual,Obrero,Ouvrier,0,2,"99, rue du Départ",NaN,1 (11) 500 555-0144,2013-03-27,0-1 Miles
2,14841,171,AW00014841,NaN,Franklin,NaN,Pal,0,1917-06-05,M,NaN,M,franklin30@adventure-works.com,20000,2,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,1,1,Wertheimer Straße 8,NaN,1 (11) 500 555-0185,2013-09-02,1-2 Miles
3,18034,160,AW00018034,NaN,Ethan,J,Robinson,0,1917-09-20,M,NaN,M,ethan34@adventure-works.com,10000,4,0,High School,Educación secundaria,Bac + 2,Manual,Obrero,Ouvrier,0,2,Celler Weg 5,NaN,1 (11) 500 555-0160,2013-05-21,0-1 Miles
4,14992,228,AW00014992,NaN,Dale,B,Andersen,0,1918-02-11,M,NaN,M,dale10@adventure-works.com,20000,2,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,1,1,2844 Plymouth Drive,NaN,1 (11) 500 555-0182,2013-09-04,5-10 Miles


In [524]:
Sales_Territory

,SalesTerritoryKey,SalesTerritoryAlternateKey,SalesTerritoryRegion,SalesTerritoryCountry,SalesTerritoryGroup
0,1,1,Northwest,United States,North America
1,2,2,Northeast,United States,North America
2,3,3,Central,United States,North America
3,4,4,Southwest,United States,North America
4,5,5,Southeast,United States,North America
5,6,6,Canada,Canada,North America
6,7,7,France,France,Europe
7,8,8,Germany,Germany,Europe
8,9,9,Australia,Australia,Pacific
9,10,10,United Kingdom,United Kingdom,Europe


## fixing datatypes, renaming cols, merging

I will go through each dataset one by one

## Products

In [525]:
#Products
Products['StartDate'] = pd.to_datetime(Products['StartDate'])
Products['EndDate'] = pd.to_datetime(Products['EndDate'])

In [526]:
psbc = Products['ProductSubcategoryKey']
psbc [psbc > 0]

209    14.0
210    14.0
211    31.0
212    31.0
213    31.0
       ... 
601     5.0
602     5.0
603     2.0
604     2.0
605     2.0
Name: ProductSubcategoryKey, Length: 397, dtype: float64

registered as float, all values are int whether > 0, >10, >n

In [527]:
Products['ProductSubcategoryKey'] = Products['ProductSubcategoryKey'].astype('Int64')

In [528]:
#Nan in langs translation cols is not translated, not an error or forgotten to be filled
translation_cols = ['SpanishProductName', 'FrenchProductName', 'EnglishDescription',
                     'FrenchDescription', 'ChineseDescription', 'ArabicDescription',
                     'HebrewDescription', 'ThaiDescription', 'GermanDescription',
                     'JapaneseDescription', 'TurkishDescription'] #ai (too lazy to write the whole list)

Products[translation_cols] = Products[translation_cols].fillna('Not Translated')

In [529]:
Products[['EndDate','Status']][250:271]

,EndDate,Status
250,2008-12-27,NaN
251,NaT,Current
252,2007-12-28,NaN
253,2008-12-27,NaN
254,NaT,Current
255,2007-12-28,NaN
256,2008-12-27,NaN
257,NaT,Current
258,2007-12-28,NaN
259,2008-12-27,NaN


alright so here we are observing a repetitive sequence on nan with a date and NaT with a current status

Products with no end date have a current status
while other with enddate have none, i think the nan's aren't random or errors but instead discontinuity of that same product

In [530]:
Products['Status'] = Products['Status'].fillna('Discontinued')

In [531]:
Products[['EndDate','Status']][250:271]

,EndDate,Status
250,2008-12-27,Discontinued
251,NaT,Current
252,2007-12-28,Discontinued
253,2008-12-27,Discontinued
254,NaT,Current
255,2007-12-28,Discontinued
256,2008-12-27,Discontinued
257,NaT,Current
258,2007-12-28,Discontinued
259,2008-12-27,Discontinued


In [532]:
Products.head()

,ProductKey,ProductAlternateKey,ProductSubcategoryKey,WeightUnitMeasureCode,SizeUnitMeasureCode,EnglishProductName,SpanishProductName,FrenchProductName,StandardCost,FinishedGoodsFlag,Color,SafetyStockLevel,ReorderPoint,ListPrice,Size,SizeRange,Weight,DaysToManufacture,ProductLine,DealerPrice,Class,Style,ModelName,EnglishDescription,FrenchDescription,ChineseDescription,ArabicDescription,HebrewDescription,ThaiDescription,GermanDescription,JapaneseDescription,TurkishDescription,StartDate,EndDate,Status
0,1,AR-5381,<NA>,NaN,NaN,Adjustable Race,Not Translated,Not Translated,NaN,0,NaN,1000,750,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,2003-07-01,NaT,Current
1,2,BA-8327,<NA>,NaN,NaN,Bearing Ball,Not Translated,Not Translated,NaN,0,NaN,1000,750,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,2003-07-01,NaT,Current
2,3,BE-2349,<NA>,NaN,NaN,BB Ball Bearing,Not Translated,Not Translated,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,2003-07-01,NaT,Current
3,4,BE-2908,<NA>,NaN,NaN,Headset Ball Bearings,Not Translated,Not Translated,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,0,NaN,NaN,NaN,NaN,NaN,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,2003-07-01,NaT,Current
4,5,BL-2036,<NA>,NaN,NaN,Blade,Not Translated,Not Translated,NaN,0,NaN,800,600,NaN,NaN,NaN,NaN,1,NaN,NaN,NaN,NaN,NaN,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,Not Translated,2003-07-01,NaT,Current


## Customers

in this dataset 3 cols gained my attention, the ones with almost 100% of nan's

In [533]:
is_na = Customers.isna().sum() / len(Customers)
is_na [is_na > 0.50]

Title           0.994536
Suffix          0.999838
AddressLine2    0.983121
dtype: float64

In [534]:
Customers[['Title', 'Suffix', 'AddressLine2']].dropna(how='all') #ai

,Title,Suffix,AddressLine2
16,NaN,NaN,# 2
40,NaN,NaN,Kreditorenbuchhaltung
120,NaN,NaN,#501
184,NaN,NaN,# 215
234,NaN,NaN,Verkaufsabteilung
...,...,...,...
18362,NaN,NaN,# 1
18412,Mr.,NaN,NaN
18415,NaN,NaN,Space 55
18416,NaN,NaN,Space 55


with 99.5% and  99.9% of NA? not only, but it's a customer dataset, title doesn't any meaning heren same for suffix
while adressline is the residency of that same customer, our dataset do have addressline1 in which it'll be mostly filled, since most clients would have at least one adresseline
so the remaining na will be filled instead of droped with empty space or None 

In [535]:
Customers = Customers.drop(columns=['Title', 'Suffix'])

Customers['AddressLine2'] = Customers['AddressLine2'].fillna('None')

In [536]:
Customers.head(3)

,CustomerKey,GeographyKey,CustomerAlternateKey,FirstName,MiddleName,LastName,NameStyle,BirthDate,MaritalStatus,Gender,EmailAddress,YearlyIncome,TotalChildren,NumberChildrenAtHome,EnglishEducation,SpanishEducation,FrenchEducation,EnglishOccupation,SpanishOccupation,FrenchOccupation,HouseOwnerFlag,NumberCarsOwned,AddressLine1,AddressLine2,Phone,DateFirstPurchase,CommuteDistance
0,12725,171,AW00012725,Gabrielle,NaN,James,0,1916-02-10,M,F,gabrielle17@adventure-works.com,10000,4,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,0,2,Nonnendamm 6500,None,1 (11) 500 555-0125,2013-08-18,1-2 Miles
1,16455,208,AW00016455,Deb,A,Moreno,0,1917-02-09,M,F,deb3@adventure-works.com,10000,4,0,High School,Educación secundaria,Bac + 2,Manual,Obrero,Ouvrier,0,2,"99, rue du Départ",None,1 (11) 500 555-0144,2013-03-27,0-1 Miles
2,14841,171,AW00014841,Franklin,NaN,Pal,0,1917-06-05,M,M,franklin30@adventure-works.com,20000,2,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,1,1,Wertheimer Straße 8,None,1 (11) 500 555-0185,2013-09-02,1-2 Miles


#Same thing for Middle name, it's absence just means the customer middle :D 

In [537]:
Customers['MiddleName'] = Customers['MiddleName'].fillna('None')

In [538]:
Customers.head(1)

,CustomerKey,GeographyKey,CustomerAlternateKey,FirstName,MiddleName,LastName,NameStyle,BirthDate,MaritalStatus,Gender,EmailAddress,YearlyIncome,TotalChildren,NumberChildrenAtHome,EnglishEducation,SpanishEducation,FrenchEducation,EnglishOccupation,SpanishOccupation,FrenchOccupation,HouseOwnerFlag,NumberCarsOwned,AddressLine1,AddressLine2,Phone,DateFirstPurchase,CommuteDistance
0,12725,171,AW00012725,Gabrielle,None,James,0,1916-02-10,M,F,gabrielle17@adventure-works.com,10000,4,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,0,2,Nonnendamm 6500,None,1 (11) 500 555-0125,2013-08-18,1-2 Miles


In [539]:
Customers[['BirthDate', 'DateFirstPurchase']]

,BirthDate,DateFirstPurchase
0,1916-02-10,2013-08-18
1,1917-02-09,2013-03-27
2,1917-06-05,2013-09-02
3,1917-09-20,2013-05-21
4,1918-02-11,2013-09-04
...,...,...
18479,1986-06-22,2013-07-08
18480,1986-06-23,2013-03-28
18481,1986-06-23,2012-05-01
18482,1986-06-25,2013-02-03


matching date frame, we only convert them

In [540]:
Customers ['BirthDate'] = pd.to_datetime(Customers['BirthDate']) 
Customers ['DateFirstPurchase'] = pd.to_datetime(Customers['DateFirstPurchase']) 

In [541]:
# Customers[Customers['Gender'] == 'm']
# Customers[Customers['Gender'] == 'f']

# Customers[Customers['MaritalStatus'] == 'm']
# Customers[Customers['MaritalStatus'] == 's']

#None for all

In [542]:
#making married_sit and gender more clearer
gender_map = {
    'M': 'Male',
    'F': 'Female',
}

marital_map = {
    'M': 'Married',
    'S': 'Single'
}

Customers['Gender'] = Customers['Gender'].map(gender_map)
Customers['MaritalStatus'] = Customers['MaritalStatus'].map(marital_map) 

In [543]:
Customers.head(3)

,CustomerKey,GeographyKey,CustomerAlternateKey,FirstName,MiddleName,LastName,NameStyle,BirthDate,MaritalStatus,Gender,EmailAddress,YearlyIncome,TotalChildren,NumberChildrenAtHome,EnglishEducation,SpanishEducation,FrenchEducation,EnglishOccupation,SpanishOccupation,FrenchOccupation,HouseOwnerFlag,NumberCarsOwned,AddressLine1,AddressLine2,Phone,DateFirstPurchase,CommuteDistance
0,12725,171,AW00012725,Gabrielle,None,James,0,1916-02-10,Married,Female,gabrielle17@adventure-works.com,10000,4,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,0,2,Nonnendamm 6500,None,1 (11) 500 555-0125,2013-08-18,1-2 Miles
1,16455,208,AW00016455,Deb,A,Moreno,0,1917-02-09,Married,Female,deb3@adventure-works.com,10000,4,0,High School,Educación secundaria,Bac + 2,Manual,Obrero,Ouvrier,0,2,"99, rue du Départ",None,1 (11) 500 555-0144,2013-03-27,0-1 Miles
2,14841,171,AW00014841,Franklin,None,Pal,0,1917-06-05,Married,Male,franklin30@adventure-works.com,20000,2,0,Partial High School,Educación secundaria (en curso),Niveau bac,Manual,Obrero,Ouvrier,1,1,Wertheimer Straße 8,None,1 (11) 500 555-0185,2013-09-02,1-2 Miles


## Sales Territory

In [544]:
# Sales_Territory
# SalesTerritoryKey	SalesTerritoryAlternateKey
# both of theses don't give us insight on what would be the territory region, or group or country
#so set to Unknown

In [545]:
Sales_Territory = Sales_Territory.fillna('Unknown')

In [546]:
Sales_Territory

,SalesTerritoryKey,SalesTerritoryAlternateKey,SalesTerritoryRegion,SalesTerritoryCountry,SalesTerritoryGroup
0,1,1,Northwest,United States,North America
1,2,2,Northeast,United States,North America
2,3,3,Central,United States,North America
3,4,4,Southwest,United States,North America
4,5,5,Southeast,United States,North America
5,6,6,Canada,Canada,North America
6,7,7,France,France,Europe
7,8,8,Germany,Germany,Europe
8,9,9,Australia,Australia,Pacific
9,10,10,United Kingdom,United Kingdom,Europe


## Internet Sales

the only issue with iternet sales is the currency code, it contains no other nans

In [547]:
#since we have the code of each number we can structure it

Currency = pd.DataFrame({
    'Key'      :          [6,     19,    29,    36,   98,   100],
    'Code'     :          ['AUD', 'CAD', 'DEM', 'EUR', 'GBP', 'USD'],
    'Currency' :          ['Australian Dollar', 'Canadian Dollar', 'Deutsche Mark',
                               'Euro', 'UK Pound', 'US Dollar'],
})

In [548]:
# Internet_Sales = Internet_Sales.drop(columns=[6,     19,    29,    36,   98,   100])

In [549]:
# Internet_Sales['Key'] = np.nan

In [550]:
# Internet_Sales.merge(Currency, on='Key')

failure attempts

In [551]:
cols = [29, 6, 98, 19, 36, 100]
Internet_Sales['Key'] = np.nan

for c in cols: #ai
    Internet_Sales.loc[Internet_Sales[c].isna() == False, 'Key'] = c

Internet_Sales = Internet_Sales.drop(columns=cols)
Internet_Sales = Internet_Sales.merge(Currency, on='Key', how='left')


In [552]:
Internet_Sales

,ProductKey,OrderDate,ShipDate,CustomerKey,PromotionKey,SalesTerritoryKey,SalesOrderNumber,SalesOrderLineNumber,RevisionNumber,OrderQuantity,UnitPrice,ExtendedAmount,UnitPriceDiscountPct,DiscountAmount,ProductStandardCost,TotalProductCost,Key,Code,Currency
0,310,2010-12-29,2011-01-05,21768,1,6,SO43697,1,1,1,3578.2700,3578.2700,0,0,2171.2942,2171.2942,19.0,CAD,Canadian Dollar
1,346,2010-12-29,2011-01-05,28389,1,7,SO43698,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,NaN,NaN,NaN
2,346,2010-12-29,2011-01-05,25863,1,1,SO43699,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,100.0,USD,US Dollar
3,336,2010-12-29,2011-01-05,14501,1,4,SO43700,1,1,1,699.0982,699.0982,0,0,413.1463,413.1463,100.0,USD,US Dollar
4,346,2010-12-29,2011-01-05,11003,1,9,SO43701,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,6.0,AUD,Australian Dollar
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
60393,485,2014-01-28,2014-02-04,15868,1,6,SO75122,1,1,1,21.9800,21.9800,0,0,8.2205,8.2205,100.0,USD,US Dollar
60394,225,2014-01-28,2014-02-04,15868,1,6,SO75122,2,1,1,8.9900,8.9900,0,0,6.9223,6.9223,100.0,USD,US Dollar
60395,485,2014-01-28,2014-02-04,18759,1,6,SO75123,1,1,1,21.9800,21.9800,0,0,8.2205,8.2205,100.0,USD,US Dollar
60396,486,2014-01-28,2014-02-04,18759,1,6,SO75123,2,1,1,159.0000,159.0000,0,0,59.4660,59.4660,100.0,USD,US Dollar


In [553]:
Internet_Sales.isnull().sum()

ProductKey               0
OrderDate                0
ShipDate                 0
CustomerKey              0
PromotionKey             0
SalesTerritoryKey        0
SalesOrderNumber         0
SalesOrderLineNumber     0
RevisionNumber           0
OrderQuantity            0
UnitPrice                0
ExtendedAmount           0
UnitPriceDiscountPct     0
DiscountAmount           0
ProductStandardCost      0
TotalProductCost         0
Key                     59
Code                    59
Currency                59
dtype: int64

In [554]:
# now that every dataset is ready, we must merge them all into one sales dataset

In [555]:
Sales = (
    Internet_Sales.merge(
        Products, on='ProductKey')
    .merge(Customers, on='CustomerKey')
    .merge(Sales_Territory, on='SalesTerritoryKey')
)

Sales.to_parquet('Sales.parquet', index=False)

In [556]:
Sales

,ProductKey,OrderDate,ShipDate,CustomerKey,PromotionKey,SalesTerritoryKey,SalesOrderNumber,SalesOrderLineNumber,RevisionNumber,OrderQuantity,UnitPrice,ExtendedAmount,UnitPriceDiscountPct,DiscountAmount,ProductStandardCost,TotalProductCost,Key,Code,Currency,ProductAlternateKey,ProductSubcategoryKey,WeightUnitMeasureCode,SizeUnitMeasureCode,EnglishProductName,SpanishProductName,FrenchProductName,StandardCost,FinishedGoodsFlag,Color,SafetyStockLevel,ReorderPoint,ListPrice,Size,SizeRange,Weight,DaysToManufacture,ProductLine,DealerPrice,Class,Style,ModelName,EnglishDescription,FrenchDescription,ChineseDescription,ArabicDescription,HebrewDescription,ThaiDescription,GermanDescription,JapaneseDescription,TurkishDescription,StartDate,EndDate,Status,GeographyKey,CustomerAlternateKey,FirstName,MiddleName,LastName,NameStyle,BirthDate,MaritalStatus,Gender,EmailAddress,YearlyIncome,TotalChildren,NumberChildrenAtHome,EnglishEducation,SpanishEducation,FrenchEducation,EnglishOccupation,SpanishOccupation,FrenchOccupation,HouseOwnerFlag,NumberCarsOwned,AddressLine1,AddressLine2,Phone,DateFirstPurchase,CommuteDistance,SalesTerritoryAlternateKey,SalesTerritoryRegion,SalesTerritoryCountry,SalesTerritoryGroup
0,310,2010-12-29,2011-01-05,21768,1,6,SO43697,1,1,1,3578.2700,3578.2700,0,0,2171.2942,2171.2942,19.0,CAD,Canadian Dollar,BK-R93R-62,2,LB,CM,"Road-150 Red, 62","Carretera: 150, roja, 62","Vélo de route 150 rouge, 62",2171.2942,1,Red,100,75,3578.2700,62,60-62 CM,15.00,4,R,2146.9620,H,U,Road-150,This bike is ridden by race winners. Developed...,Ce vélo est destiné aux champions cyclistes. M...,此自行车曾为赛车手夺冠立下汗马功劳。与 Adventure Works Cycles 专业赛...,إنها الدراجة التي يقودها الفائزون بسباقات الدر...,אופניים אלה משמשים את מנצחי התחרויות. לאופניים...,จักรยานคู่ใจของนักแข่งระดับแชมเปี้ยน ได้รับกา...,Dieses Fahrrad wird von Rennprofis gefahren. E...,レース ウィナーも愛用する機種です。Adventure Works Cycles プロフェッ...,Bu bisiklet yarışçılar tarafından kullanılmakt...,2011-07-01,2007-12-28,Discontinued,53,AW00021768,Cole,A,Watson,0,1952-02-19,Single,Male,cole1@adventure-works.com,70000,5,0,Bachelors,Licenciatura,Bac + 4,Management,Gestión,Direction,1,3,601 Asilomar Dr.,None,110-555-0129,2010-12-29,10+ Miles,6,Canada,Canada,North America
1,346,2010-12-29,2011-01-05,28389,1,7,SO43698,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,NaN,NaN,NaN,BK-M82S-44,1,LB,CM,"Mountain-100 Silver, 44","Montaña: 100, plateada, 44","VTT 100 argent, 44",1912.1544,1,Silver,100,75,3399.9900,44,42-46 CM,21.13,4,M,2039.9940,H,U,Mountain-100,Top-of-the-line competition mountain bike. Per...,VTT de compétition haut de gamme. Plusieurs op...,高档竞赛山地车。性能得到进一步增强，包括创新的 HL 车架、极其平稳的前悬架以及适用于所有地...,دراجة سباقات مخصصة للقيادة في الجبال من أعلى ط...,אופני הרים תחרותיים מהשורה הראשונה. אפשרויות ל...,สุดยอดจักรยานภูเขาสำหรับการแข่งขัน สมบูรณ์แบบด...,Top-Mountainbike für Profis. Mit innovativem H...,最高級の競技用マウンテン バイクです。性能強化オプションとしては、革新的な HL フレーム、...,En üst model rekabetçi dağ bisikleti. Performa...,2011-07-01,2007-12-28,Discontinued,222,AW00028389,Rachael,M,Martinez,0,1970-06-17,Single,Female,rachael16@adventure-works.com,20000,3,3,High School,Educación secundaria,Bac + 2,Manual,Obrero,Ouvrier,0,0,"14, avenue du Port",None,1 (11) 500 555-0190,2010-12-29,0-1 Miles,7,France,France,Europe
2,346,2010-12-29,2011-01-05,25863,1,1,SO43699,1,1,1,3399.9900,3399.9900,0,0,1912.1544,1912.1544,100.0,USD,US Dollar,BK-M82S-44,1,LB,CM,"Mountain-100 Silver, 44","Montaña: 100, plateada, 44","VTT 100 argent, 44",1912.1544,1,Silver,100,75,3399.9900,44,42-46 CM,21.13,4,M,2039.9940,H,U,Mountain-100,Top-of-the-line competition mountain bike. Per...,VTT de compétition haut de gamme. Plusieurs op...,高档竞赛山地车。性能得到进一步增强，包括创新的 HL 车架、极其平稳的前悬架以及适用于所有地...,دراجة سباقات مخصصة للقيادة في الجبال من أعلى ط...,אופני הרים תחרותיים מהשורה הראשונה. אפשרויות ל...,สุดยอดจักรยานภูเขาสำหรับการแข่งขัน สมบูรณ์แบบด...,Top-Mountainbike für Profis. Mit innovativem H...,最高級の競技用マウンテン バイクです。性能強化オプションとしては、革新的な HL フレーム、...,En üst model rekabetçi dağ bisikleti. Performa...,